File: src/edgar.ts
import { FundamentalData, type PERatioData } from './types';

const SEC_COMPANY_TICKERS_URL = 'https://www.sec.gov/files/company_tickers.json';
const SEC_DATA_BASE_URL = 'https://data.sec.gov';

// SEC requires a declared User-Agent with a contact. A URL in the User-Agent
// is rejected, so keep it to name + contact. Override via SEC_USER_AGENT.
const DEFAULT_SEC_USER_AGENT = 'stock-fetcher/1.0 (admin@example.com)';

// SEC fair-access limit is 10 requests/second. Serializing at 5/second keeps
// comfortably below the threshold.
const SEC_MIN_REQUEST_INTERVAL_MS = 200;

export interface EdgarEPSFact {
  start: string;
  end: string;
  val: number;
  form: string;
  fp: string;
  filed: string;
  frame?: string;
}

interface EdgarCompanyFacts {
  entityName: string;
  facts: {
    'us-gaap': {
      EarningsPerShareDiluted: {
        units: Record<string, EdgarEPSFact[]>;
      };
    };
  };
}

// Per-isolate cache for the ~800KB SEC ticker map.
let tickersCache: Record<string, string> | null = null;
let tickersInFlight: Promise<Record<string, string>> | null = null;

let lastRequestTime = 0;
let requestQueue: Promise<unknown> = Promise.resolve();

let configuredUserAgent = DEFAULT_SEC_USER_AGENT;

// Allows the request handler to override the SEC User-Agent from env.
export function setSECUserAgent(value: string | undefined): void {
  configuredUserAgent = value || DEFAULT_SEC_USER_AGENT;
}

function userAgent(): string {
  return configuredUserAgent;
}

// Performs a serialized, rate-limited SEC request.
async function secFetch(url: string): Promise<Response> {
  const run = async (): Promise<Response> => {
    const now = Date.now();
    const wait = Math.max(0, SEC_MIN_REQUEST_INTERVAL_MS - (now - lastRequestTime));
    if (wait > 0) {
      await new Promise((resolve) => setTimeout(resolve, wait));
    }
    lastRequestTime = Date.now();
    return fetch(url, {
      headers: {
        'User-Agent': userAgent(),
        Accept: 'application/json',
      },
    });
  };

  const result = requestQueue.then(run, run);
  requestQueue = result.then(
    () => undefined,
    () => undefined,
  );
  return result;
}

async function readJSON<T>(resp: Response, context: string): Promise<T> {
  if (!resp.ok) {
    const body = (await resp.text()).slice(0, 512);
    throw new Error(`SEC returned status ${resp.status}: ${body}`);
  }
  try {
    return (await resp.json()) as T;
  } catch (err) {
    throw new Error(`failed to parse SEC response (${context}): ${err}`);
  }
}

// Resolves a ticker to its zero-padded 10-digit CIK.
export async function getCIK(symbol: string): Promise<string> {
  const upper = symbol.toUpperCase();

  if (tickersCache?.[upper]) return tickersCache[upper];

  if (!tickersInFlight) {
    tickersInFlight = (async () => {
      const resp = await secFetch(SEC_COMPANY_TICKERS_URL);
      const data = await readJSON<Record<string, { cik_str: number; ticker: string }>>(
        resp,
        'ticker map',
      );

      const map: Record<string, string> = {};
      for (const value of Object.values(data)) {
        if (value.ticker) {
          map[value.ticker.toUpperCase()] = String(value.cik_str).padStart(10, '0');
        }
      }
      tickersCache = map;
      return map;
    })().finally(() => {
      tickersInFlight = null;
    });
  }

  const map = await tickersInFlight;
  const cik = map[upper];
  if (!cik) {
    throw new Error(`symbol ${symbol} not found in SEC ticker map`);
  }
  return cik;
}

// Fetches historical TTM EPS from EDGAR in the same FundamentalData shape
// used by the macrotrends fetcher.
export async function fetchEDGARFundamental(symbol: string): Promise<FundamentalData> {
  const cik = await getCIK(symbol);
  const factsURL = `${SEC_DATA_BASE_URL}/api/xbrl/companyfacts/CIK${cik}.json`;

  const resp = await secFetch(factsURL);
  const facts = await readJSON<EdgarCompanyFacts>(resp, 'company facts');

  const rawFacts = facts.facts['us-gaap'].EarningsPerShareDiluted.units['USD/shares'] ?? [];
  const quarters = selectQuarterlyEPS(rawFacts);
  if (quarters.length === 0) {
    throw new Error(`no quarterly diluted EPS data found for ${symbol}`);
  }

  const withQ4 = deriveMissingQ4(quarters, selectAnnualEPS(rawFacts));
  const history = buildTTMEPSSeries(withQ4);

  return new FundamentalData(
    symbol.toUpperCase(),
    facts.entityName,
    history,
    0,
    latestPositiveEPS(history),
    0,
  );
}

// Selects one diluted EPS value per fiscal quarter from raw XBRL facts.
export function selectQuarterlyEPS(facts: EdgarEPSFact[]): EdgarEPSFact[] {
  const best = new Map<string, EdgarEPSFact>();

  for (const fact of facts) {
    if (!fact.start || !fact.end) continue;
    if (fact.form !== '10-Q' && fact.form !== '10-K' && fact.form !== '10-K/A') continue;

    const duration = daysBetween(fact.start, fact.end);
    if (duration <= 0 || duration >= 120) continue;

    const current = best.get(fact.end);
    if (
      !current ||
      fact.filed > current.filed ||
      (fact.filed === current.filed && duration < daysBetween(current.start, current.end))
    ) {
      best.set(fact.end, fact);
    }
  }

  return [...best.values()].sort((a, b) => (a.end < b.end ? -1 : a.end > b.end ? 1 : 0));
}

// Selects one diluted EPS value per fiscal year from 10-K/10-K/A facts.
export function selectAnnualEPS(facts: EdgarEPSFact[]): EdgarEPSFact[] {
  const best = new Map<string, EdgarEPSFact>();

  for (const fact of facts) {
    if (!fact.start || !fact.end) continue;
    if (fact.form !== '10-K' && fact.form !== '10-K/A') continue;

    const duration = daysBetween(fact.start, fact.end);
    if (duration < 300) continue;

    const current = best.get(fact.end);
    if (!current || fact.filed > current.filed) {
      best.set(fact.end, fact);
    }
  }

  return [...best.values()].sort((a, b) => (a.end < b.end ? -1 : a.end > b.end ? 1 : 0));
}

// Fills in the fourth fiscal quarter for years where only annual EPS was
// tagged in the 10-K: Q4 = annual EPS - the three available quarters.
export function deriveMissingQ4(quarters: EdgarEPSFact[], annuals: EdgarEPSFact[]): EdgarEPSFact[] {
  const covered = new Set(quarters.map((q) => q.end));
  const result = [...quarters];

  for (const annual of annuals) {
    if (covered.has(annual.end)) continue;

    const candidates = quarters.filter((q) => {
      const daysBefore = daysBetween(q.end, annual.end);
      return daysBefore > 0 && daysBefore < 320;
    });

    if (candidates.length < 3) continue;

    const last3 = candidates.slice(-3);
    const q4Value = annual.val - last3[0].val - last3[1].val - last3[2].val;

    result.push({
      start: last3[2].end,
      end: annual.end,
      val: q4Value,
      form: annual.form,
      fp: 'Q4',
      filed: annual.filed,
    });
  }

  return result.sort((a, b) => (a.end < b.end ? -1 : a.end > b.end ? 1 : 0));
}

// Builds trailing-twelve-month EPS data points from quarterly EPS facts.
export function buildTTMEPSSeries(quarters: EdgarEPSFact[]): PERatioData[] {
  const series: PERatioData[] = [];

  for (let i = 0; i < quarters.length; i++) {
    const quarter = quarters[i];
    let eps = 0;
    if (i >= 3) {
      eps = quarters[i - 3].val + quarters[i - 2].val + quarters[i - 1].val + quarter.val;
    }

    series.push({
      date: quarter.end,
      stockPrice: 0,
      eps,
      peRatio: 0,
    });
  }

  return series;
}

// Returns the number of days between two YYYY-MM-DD dates, or 0 on error.
export function daysBetween(start: string, end: string): number {
  const startMs = Date.parse(start + 'T00:00:00Z');
  const endMs = Date.parse(end + 'T00:00:00Z');
  if (Number.isNaN(startMs) || Number.isNaN(endMs)) return 0;
  return Math.round((endMs - startMs) / 86_400_000);
}

// Returns the most recent positive EPS value, or 0 if none exists.
export function latestPositiveEPS(history: PERatioData[]): number {
  for (let i = history.length - 1; i >= 0; i--) {
    if (history[i].eps > 0) return history[i].eps;
  }
  return 0;
}

File: edgar.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"sort"
	"strings"
	"sync"
	"time"
)

const (
	secCompanyTickersURL = "https://www.sec.gov/files/company_tickers.json"
	secDataBaseURL       = "https://data.sec.gov"

	// SEC requires a declared, non-browser User-Agent (see sec.gov/os/accessing-edgar-data).
	// A URL in the User-Agent is rejected, so keep it to name + contact.
	secUserAgent = "stock-fetcher/1.0 (admin@example.com)"

	// SEC fair-access limit is 10 requests/second. Serializing at 5/second
	// keeps comfortably below the threshold.
	secMinRequestInterval = 200 * time.Millisecond
)

// EDGARFetcher fetches historical EPS data from the SEC EDGAR XBRL API.
// EDGAR is free, official, and requires no API key.
//
// It does not provide market prices, so callers combine EDGAR fundamentals
// with prices from another source (e.g. Yahoo Finance).
type EDGARFetcher struct {
	client *http.Client

	// Test seam for swapping the upstream URLs.
	tickersURL   string
	factsBaseURL string
	userAgent    string

	// tickers caches the ticker -> CIK map after the first lookup.
	tickerMu sync.Mutex
	tickers  map[string]string

	// rateMu serializes SEC requests and enforces the rate interval.
	rateMu  sync.Mutex
	lastReq time.Time
}

// NewEDGARFetcher creates a new SEC EDGAR fetcher.
// Set SEC_USER_AGENT to override the User-Agent sent to SEC (SEC requires a
// declared contact, e.g. "StockFetcher/1.0 (you@example.com)").
func NewEDGARFetcher() *EDGARFetcher {
	userAgent := os.Getenv("SEC_USER_AGENT")
	if userAgent == "" {
		userAgent = secUserAgent
	}

	return &EDGARFetcher{
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
		tickersURL:   secCompanyTickersURL,
		factsBaseURL: secDataBaseURL,
		userAgent:    userAgent,
		tickers:      make(map[string]string),
	}
}

// EdgarEPSFact is a single XBRL fact for us-gaap:EarningsPerShareDiluted.
type EdgarEPSFact struct {
	Start string  `json:"start"`
	End   string  `json:"end"`
	Val   float64 `json:"val"`
	Form  string  `json:"form"`
	FP    string  `json:"fp"`
	Filed string  `json:"filed"`
	Frame string  `json:"frame"`
}

// sharedEDGARFetcher is used by the fallback path so the ~800KB SEC ticker
// map is only downloaded once per process instead of once per request.
var sharedEDGARFetcher = NewEDGARFetcher()

// EdgarCompanyFacts is the subset of the EDGAR companyfacts JSON we need.
type EdgarCompanyFacts struct {
	EntityName string `json:"entityName"`
	Facts      struct {
		UsGAAP struct {
			EarningsPerShareDiluted struct {
				Units map[string][]EdgarEPSFact `json:"units"`
			} `json:"EarningsPerShareDiluted"`
		} `json:"us-gaap"`
	} `json:"facts"`
}

// fetchJSON performs a rate-limited GET request and decodes the JSON body.
func (f *EDGARFetcher) fetchJSON(url string, out interface{}) error {
	f.rateMu.Lock()
	defer f.rateMu.Unlock()

	if wait := secMinRequestInterval - time.Since(f.lastReq); wait > 0 {
		time.Sleep(wait)
	}
	f.lastReq = time.Now()

	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return err
	}
	req.Header.Set("User-Agent", f.userAgent)
	req.Header.Set("Accept", "application/json")

	resp, err := f.client.Do(req)
	if err != nil {
		return err
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 512))
		return fmt.Errorf("SEC returned status %d: %s", resp.StatusCode, strings.TrimSpace(string(body)))
	}

	if err := json.NewDecoder(resp.Body).Decode(out); err != nil {
		return fmt.Errorf("failed to parse SEC response: %w", err)
	}
	return nil
}

// getCIK resolves a ticker symbol to its zero-padded 10-digit CIK.
func (f *EDGARFetcher) getCIK(symbol string) (string, error) {
	upper := strings.ToUpper(symbol)

	f.tickerMu.Lock()
	defer f.tickerMu.Unlock()

	if cik, ok := f.tickers[upper]; ok {
		return cik, nil
	}

	var data map[string]struct {
		CIK    int    `json:"cik_str"`
		Ticker string `json:"ticker"`
	}
	if err := f.fetchJSON(f.tickersURL, &data); err != nil {
		return "", fmt.Errorf("failed to load SEC ticker map: %w", err)
	}

	for _, v := range data {
		if v.Ticker != "" {
			f.tickers[strings.ToUpper(v.Ticker)] = fmt.Sprintf("%010d", v.CIK)
		}
	}

	cik, ok := f.tickers[upper]
	if !ok {
		return "", fmt.Errorf("symbol %s not found in SEC ticker map", symbol)
	}
	return cik, nil
}

// FetchFundamental fetches historical TTM EPS data from EDGAR and returns it
// in the same FundamentalData shape used by the macrotrends fetcher.
// StockPrice and PERatio are left zero because EDGAR has no market prices.
func (f *EDGARFetcher) FetchFundamental(symbol string) (*FundamentalData, error) {
	cik, err := f.getCIK(symbol)
	if err != nil {
		return nil, err
	}

	factsURL := fmt.Sprintf("%s/api/xbrl/companyfacts/CIK%s.json", f.factsBaseURL, cik)
	var facts EdgarCompanyFacts
	if err := f.fetchJSON(factsURL, &facts); err != nil {
		return nil, fmt.Errorf("failed to fetch SEC company facts: %w", err)
	}

	rawFacts := facts.Facts.UsGAAP.EarningsPerShareDiluted.Units["USD/shares"]
	quarters := selectQuarterlyEPS(rawFacts)
	if len(quarters) == 0 {
		return nil, fmt.Errorf("no quarterly diluted EPS data found for %s", symbol)
	}

	quarters = deriveMissingQ4(quarters, selectAnnualEPS(rawFacts))
	history := buildTTMEPSSeries(quarters)

	return &FundamentalData{
		Symbol:         strings.ToUpper(symbol),
		CompanyName:    facts.EntityName,
		CurrentEPS:     latestPositiveEPS(history),
		HistoricalData: history,
	}, nil
}

// selectQuarterlyEPS reduces raw EDGAR facts to one diluted EPS value per
// fiscal quarter. It keeps facts whose start->end duration looks like a
// single quarter (under 120 days), discarding year-to-date and annual facts.
// When the same quarter appears in multiple filings (original + amended),
// the most recently filed value wins.
func selectQuarterlyEPS(facts []EdgarEPSFact) []EdgarEPSFact {
	best := make(map[string]EdgarEPSFact)

	for _, fact := range facts {
		if fact.Start == "" || fact.End == "" {
			continue
		}
		if fact.Form != "10-Q" && fact.Form != "10-K" && fact.Form != "10-K/A" {
			continue
		}

		duration := daysBetween(fact.Start, fact.End)
		if duration <= 0 || duration >= 120 {
			continue
		}

		current, exists := best[fact.End]
		if !exists || fact.Filed > current.Filed ||
			(fact.Filed == current.Filed && duration < daysBetween(current.Start, current.End)) {
			best[fact.End] = fact
		}
	}

	result := make([]EdgarEPSFact, 0, len(best))
	for _, fact := range best {
		result = append(result, fact)
	}

	sort.Slice(result, func(i, j int) bool {
		return result[i].End < result[j].End
	})

	return result
}

// selectAnnualEPS reduces raw EDGAR facts to one diluted EPS value per fiscal
// year. It keeps 10-K/10-K/A facts whose duration looks like a fiscal year.
func selectAnnualEPS(facts []EdgarEPSFact) []EdgarEPSFact {
	best := make(map[string]EdgarEPSFact)

	for _, fact := range facts {
		if fact.Start == "" || fact.End == "" {
			continue
		}
		if fact.Form != "10-K" && fact.Form != "10-K/A" {
			continue
		}

		duration := daysBetween(fact.Start, fact.End)
		if duration < 300 {
			continue
		}

		current, exists := best[fact.End]
		if !exists || fact.Filed > current.Filed {
			best[fact.End] = fact
		}
	}

	result := make([]EdgarEPSFact, 0, len(best))
	for _, fact := range best {
		result = append(result, fact)
	}

	sort.Slice(result, func(i, j int) bool {
		return result[i].End < result[j].End
	})

	return result
}

// deriveMissingQ4 fills in the fourth fiscal quarter for years where the
// filer only tagged annual EPS in the 10-K. The Q4 value is annual EPS minus
// the three quarterly values already available for that fiscal year.
func deriveMissingQ4(quarters, annuals []EdgarEPSFact) []EdgarEPSFact {
	covered := make(map[string]bool, len(quarters))
	for _, q := range quarters {
		covered[q.End] = true
	}

	result := append([]EdgarEPSFact(nil), quarters...)

	for _, annual := range annuals {
		if covered[annual.End] {
			continue
		}

		// Find the three most recent quarter ends that fall within the fiscal
		// year (within 320 days before the fiscal year end).
		var candidates []EdgarEPSFact
		for _, q := range quarters {
			daysBefore := daysBetween(q.End, annual.End)
			if daysBefore > 0 && daysBefore < 320 {
				candidates = append(candidates, q)
			}
		}

		if len(candidates) < 3 {
			continue
		}

		last3 := candidates[len(candidates)-3:]
		q4Value := annual.Val - last3[0].Val - last3[1].Val - last3[2].Val

		result = append(result, EdgarEPSFact{
			Start: last3[2].End,
			End:   annual.End,
			Val:   q4Value,
			Form:  annual.Form,
			FP:    "Q4",
			Filed: annual.Filed,
		})
	}

	sort.Slice(result, func(i, j int) bool {
		return result[i].End < result[j].End
	})

	return result
}

// buildTTMEPSSeries converts quarterly EPS facts into trailing-twelve-month
// EPS data points. The first three quarters have EPS 0 because a full TTM
// window is not yet available.
func buildTTMEPSSeries(quarters []EdgarEPSFact) []PERatioData {
	series := make([]PERatioData, 0, len(quarters))

	for i, quarter := range quarters {
		eps := 0.0
		if i >= 3 {
			eps = quarters[i-3].Val + quarters[i-2].Val + quarters[i-1].Val + quarter.Val
		}

		series = append(series, PERatioData{
			Date: quarter.End,
			EPS:  eps,
		})
	}

	return series
}

// daysBetween returns the number of days between two YYYY-MM-DD dates.
// It returns 0 if either date cannot be parsed.
func daysBetween(start, end string) int {
	startDate, startErr := time.Parse("2006-01-02", start)
	endDate, endErr := time.Parse("2006-01-02", end)
	if startErr != nil || endErr != nil {
		return 0
	}
	return int(endDate.Sub(startDate).Hours() / 24)
}

// latestPositiveEPS returns the most recent positive EPS value in the series,
// or 0 if none exists.
func latestPositiveEPS(history []PERatioData) float64 {
	for i := len(history) - 1; i >= 0; i-- {
		if history[i].EPS > 0 {
			return history[i].EPS
		}
	}
	return 0
}
